### ISSUE TITLE
Something wrong with Keyer interface?

### ISSUE THREAD

**Issue by hanleym** [gopherjs/vecty#185]
When changing the order of a `vecty.List` the elements don't always get reordered correctly.

To recreate the issue sort the `Category` column twice so that it is in descending order. In the console you'll see that the slice has been sorted correctly, however, the DOM elements haven't been reordered correctly.

```go
package main

import (
	"fmt"
	"sort"

	"github.com/gopherjs/vecty"
	"github.com/gopherjs/vecty/elem"
	"github.com/gopherjs/vecty/event"
)

type root struct {
	vecty.Core
	Results       []result
	sortColumn    string
	sortDirection string
}

type row struct {
	vecty.Core
	Result result
}

type result struct {
	Domain   string
	Category string
}

func main() {
	vecty.SetTitle("Vecty Sort")
	vecty.AddStylesheet("https://cdn.jsdelivr.net/npm/semantic-ui@2/dist/semantic.min.css")
	vecty.RenderBody(&root{
		Results: []result{
			result{Domain: "bing.com", Category: "Search"},
			result{Domain: "facebook.com", Category: "Social Media"},
			result{Domain: "google.com", Category: "Search"},
			result{Domain: "instagram.com", Category: "Social Media"},
			result{Domain: "myspace.com", Category: "Social Media"},
			result{Domain: "somethingawful.com", Category: "Social Media"},
			result{Domain: "twitter.com", Category: "Social Media"},
			result{Domain: "pinterest.com", Category: "Social Media"},
			result{Domain: "reddit.com", Category: "Social Media"},
			result{Domain: "snapchat.com", Category: "Social Media"},
		},
	})
}

func (r *root) Sort(column string) func(*vecty.Event) {
	return func(e *vecty.Event) {
		if r.sortColumn == column {
			if r.sortDirection == "down" {
				r.sortDirection = "up"
			} else {
				r.sortDirection = "down"
			}
		} else {
			r.sortColumn = column
			r.sortDirection = "up"
		}

		println(r.sortColumn, r.sortDirection)

		sort.Slice(r.Results, func(a, b int) bool {
			if r.sortDirection == "down" {
				a, b = b, a
			}
			switch r.sortColumn {
			case "domain":
				return r.Results[a].Domain < r.Results[b].Domain
			case "category":
				return r.Results[a].Category < r.Results[b].Category
			}
			panic("sortColumn not found")
		})

		for _, result := range r.Results {
			fmt.Printf("%#v\n", result)
		}

		vecty.Rerender(r)
	}
}

func (r *root) Render() vecty.ComponentOrHTML {
	rows := vecty.List{}
	for _, result := range r.Results {
		rows = append(rows, &row{
			Result: result,
		})
	}

	return elem.Body(
		elem.Table(
			vecty.Markup(
				vecty.Class("ui", "selectable", "striped", "unstackable", "single", "line", "table"),
				vecty.Style("margin-bottom", "20px"),
			),
			elem.TableHead(
				elem.TableRow(
					elem.TableHeader(
						vecty.Markup(
							event.Click(r.Sort("domain")),
							vecty.Style("cursor", "pointer"),
						),
						vecty.Text("Domain"),
						vecty.If(
							r.sortColumn == "domain",
							elem.Italic(
								vecty.Markup(
									vecty.Class("caret", r.sortDirection, "icon"),
								),
							),
						),
					),
					elem.TableHeader(
						vecty.Markup(
							event.Click(r.Sort("category")),
							vecty.Style("cursor", "pointer"),
						),
						vecty.Text("Category"),
						vecty.If(
							r.sortColumn == "category",
							elem.Italic(
								vecty.Markup(
									vecty.Class("caret", r.sortDirection, "icon"),
								),
							),
						),
					),
				),
			),
			elem.TableBody(
				rows,
			),
		),
	)
}

func (r *row) Key() interface{} {
	return r.Result.Domain
}

func (r *row) Render() vecty.ComponentOrHTML {
	return elem.TableRow(
		elem.TableData(
			vecty.Text(r.Result.Domain),
		),
		elem.TableData(
			vecty.Text(r.Result.Category),
		),
	)
}
```

**Comment by slimsag** [comment]
I followed up on Slack about this. Posting my response here for posterity:

> @hanleym I see two issues here. One is a bug in your code, and one looks to be a bug in Vecty (but I don't have enough time to dive into that right now to be sure)
>
> @hanleym For the bug in your code:
you have a component:
>
```go
type row struct {
    vecty.Core
    Result result
}
```
>
> this component is created with the `Result` value passed in, i.e.
```go
    for _, result := range r.Results {
        rows = append(rows, &row{
            Result: result,
        })
    }
```
>
> but when that component is rendered a second time, you're expecting that new row's result value to be different. It won't be, because Vecty considers it to be state (and therefor will e.g. copy the `oldRow.Result` value into the `newRow.Result` value)
>
> fixing this is simple, just change `Result result` to `Result result `vecty:"prop"`` and Vecty will not copy the old value of that field over when you instantiate the component again
>
> @hanleym For the bug in Vecty, there seems to be something wrong here with the Keyer interface. Removing / commenting out the `row.Key` method definition fixes the issue (but this has other drawbacks, like causing elements to be needlessly recreated, so it is not a good solution / I really need to find the bug here soon :slightly_smiling_face: )

**Comment by slimsag** [comment]
TL;DR: something wrong with Keyer interface here, try this program with and without the Keyer method commented out:

https://gist.github.com/slimsag/aad7dd558c121a883b177dd6b5d090d0#file-gistfile1-go-L139-L142

**Comment by pdf** [comment]
Hrm, weird that it works fine sorting by one column, but not the other, I'll take a look at it.

**Comment by pdf** [comment]
It's caused by the stable key code for keyed siblings (that aims to detect when a node has not changed position to avoid performing an insert).  This is a perf optimization that we could simply remove in the short term for correct behaviour, but I'll see if I can find the time to discern a proper fix in the next few days.
